### ISSUE TITLE
[Stretch] TKLBAM restore failing...

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#970]
Thanks to @DocCyblade for resolving #962. Next up is tklbam-restore issues... on face value, it looks like squid isn't playing nice, although perhaps it's related to permissions? Here's the full `tklbam-restore` output:
```
root@core ~# tklbam-restore 50
Executing Duplicity to download s3://s3-ap-southeast-2.amazonaws.com/tklbam-ap-southeast-2-34b10189773e284b420f/hloyrx7innyqteue to /tmp/tklbam-vb0qQv 
=======================================================================================================================================================

// started squid: caching downloaded backup archives to /var/cache/tklbam/restore

Traceback (most recent call last):
  File "/usr/bin/tklbam-restore", line 553, in <module>
    main()
  File "/usr/bin/tklbam-restore", line 491, in main
    backup_extract_path = get_backup_extract()
  File "/usr/bin/tklbam-restore", line 463, in get_backup_extract
    downloader(raw_download_path, target, log=_print if not silent else None, debug=opt_debug, force=opt_force)
  File "/usr/lib/tklbam/duplicity.py", line 166, in __call__
    squid.start()
  File "/usr/lib/tklbam/squid.py", line 59, in start
    raise Error("%s failed to start\n" % SQUID_BIN + self.command.output)
squid.Error: /usr/lib/tklbam/deps/usr/sbin/tklbam-squid failed to start
2017/11/20 06:19:02| Creating Swap Directories
FATAL: Failed to make swap directory /var/cache/tklbam/restore/00: (13) Permission denied
Squid Cache (Version 2.7.STABLE9): Terminated abnormally.
CPU Usage: 0.000 seconds = 0.000 user + 0.000 sys
Maximum Resident Size: 1152 KB
Page faults with physical i/o: 0
2017/11/20 06:19:02| Starting Squid Cache version 2.7.STABLE9 for x86_64-debian-linux-gnu...
2017/11/20 06:19:02| Process ID 4470
2017/11/20 06:19:02| With 1024 file descriptors available
2017/11/20 06:19:02| Using epoll for the IO loop
2017/11/20 06:19:02| DNS Socket created at 0.0.0.0, port 49511, FD 7
2017/11/20 06:19:02| Adding nameserver 192.168.1.1 from /etc/resolv.conf
2017/11/20 06:19:02| User-Agent logging is disabled.
2017/11/20 06:19:02| Referer logging is disabled.
2017/11/20 06:19:02| Unlinkd pipe opened on FD 11
2017/11/20 06:19:02| Swap maxSize 1041408 + 0 KB, estimated 80108 objects
2017/11/20 06:19:02| Target number of buckets: 4005
2017/11/20 06:19:02| Using 8192 Store buckets
2017/11/20 06:19:02| Max Mem  size: 0 KB
2017/11/20 06:19:02| Max Swap size: 1041408 KB
2017/11/20 06:19:02| Local cache digest enabled; rebuild/rewrite every 3600/3600 sec
2017/11/20 06:19:02| Store logging disabled
2017/11/20 06:19:02| /var/cache/tklbam/restore/00: (13) Permission denied
FATAL: 	Failed to verify one of the swap directories, Check cache.log
	for details.  Run 'squid -z' to create swap directories
	if needed, or if running Squid for the first time.
Aborted
Traceback (most recent call last):
  File "/usr/lib/tklbam/deps/usr/sbin/tklbam-squid", line 192, in <module>
    main()
  File "/usr/lib/tklbam/deps/usr/sbin/tklbam-squid", line 189, in main
    system("%s -f %s -N -D -d 1" % (squid_bin, conf_tmp.path))
  File "/usr/lib/python2.7/dist-packages/executil.py", line 56, in system
    raise ExecError(command, exitcode)
executil.ExecError: non-zero exitcode (134) for command: /usr/lib/tklbam/deps/usr/sbin/squid -f /tmp/tklbam-squid-conf-yzSCol -N -D -d 1
```
Then I just tried running squid directly:
```
root@core ~# /usr/lib/tklbam/deps/usr/sbin/tklbam-squid
error: incorrect number of arguments
Syntax: /usr/lib/tklbam/deps/usr/sbin/tklbam-squid <http_port> <cache_size>
Runs a squid instance to cache TKLBAM's duplicity archives
Arguments:

    <http_port>         [hostname:]port (e.g., 3128, 127.0.0.1:33128)
    <cachesize>         Size of squid cache (can be absolute or relative size)
        
                        1000MB - 1000 MBs
                        2GB - 2000 MBs

                        50%                                                  of free space in spool location

Environment variables:

    TKLBAM_SQUID_CACHE_DIR      location of cache dir (default: /var/spool/tklbam-squid)
    TKLBAM_SQUID_USER           user under which we execute squid (default: proxy)
    TKLBAM_SQUID_BIN            path to tklbam squid binary (default: /usr/lib/tklbam/deps/usr/sbin/squid)
```
The permissions appear to be right in the cache dir:
```
root@core ~# ls -l /var/cache/tklbam/
total 4
drwxr-x--- 2 proxy proxy 4096 Nov 20 02:55 restore
```
Want to have a look at this one too @DocCyblade? If so, please allocate yourself.

**Comment by DocCyblade** [comment]
So, I finally had some time to take a peek at this last night.

The error seems pretty straight forward, its an access issues with the cache directory:

`2017/11/20 06:19:02| /var/cache/tklbam/restore/00: (13) Permission denied`

The question is, why. It seemed the issue was with squid itself, as if I called the wrapper:
`./tklbam-squid 8888 200M`

I still got the error. I added the `-X` to [the command in the wrapper code](https://github.com/turnkeylinux/tklbam-squid/blob/master/tklbam-squid#L189) and looked at the output. I noticed the nobody and the cache_effective_user setting. 

```
...
2018/01/08 03:10:26| parse_line: mail_program mail
2018/01/08 03:10:26| parse_line: cache_effective_user nobody
2018/01/08 03:10:26| parse_line: httpd_suppress_version_string off
2018/01/08 03:10:26| parse_line: umask 027
...
2018/01/08 03:10:26| leave_suid: PID 3925 called
2018/01/08 03:10:26| leave_suid: PID 3925 giving up root, becoming 'nobody'
2018/01/08 03:10:26| Starting Squid Cache version 2.7.STABLE9 for x86_64-debian-linux-gnu...
2018/01/08 03:10:26| Process ID 3925
...
```

I did a quick look on squid docs on google and it seems that is the default user is nobody for that setting, so I took a look at TKL source as we patch and build our own binary. I found we do [patch it and make the default proxy](https://github.com/turnkeylinux/tklbam-squid/blob/master/debian/patches/01-cf.data.debian.patch#L139).

I am not sure why but compiled version of squid is defaulting to use the user `nobody` instead of `proxy` but that is the issue here.

A quick fix was to override the `cache_effective_user` setting to `proxy` [above this line](https://github.com/turnkeylinux/tklbam-squid/blob/master/tklbam-squid#L67) 

```python
...

cache_log /dev/null
access_log none
cache_store_log none
pid_filename none
netdb_filename none

cache_effective_user proxy
"""

class Error(Exception):
    pass
...
```

Making that change and trying a restore everything worked an it should. 

The real fix should be looking at why our patch did not patch squid.

**Comment by JedMeister** [comment]
Cheers @DocCyblade.

@lamby - do you have any ideas why our patch isn't being applied? I see that we've merged your reproducible packaging changes and that that includes changes to the patches. I'll try to have a quick look myself, in case it's just a simple typo somewhere or something, but unless I post back with anything, can you please take a look?

**Comment by lamby** [comment]
@JedMeister Did you push this out via a change to a [Debian conffile](https://wiki.debian.org/DpkgConffileHandling) ie. not *necessary* a patch to a binary? If so, the value won't end up on the target system.

**Comment by JedMeister** [comment]
@lamby - sorry for slow response...

TBH, I'm not sure. As you may recall, Debian packaging is certainly not my strong suit! :smile: 

All I know is that prior to the changes to make the [tklbam-squid](https://github.com/turnkeylinux/tklbam-squid) package reproducable (i.e. [this merge](https://github.com/turnkeylinux/tklbam-squid/commit/3b4db5d520cf52fa96f7a597e7ce99ad800ad5be)) TKLBAM worked ok. Since those changes have been merged and the package rebuilt for stretch it's no longer working (see @DocCyblade's post [above](https://github.com/turnkeylinux/tracker/issues/970#issuecomment-355941337)).

In fairness, I haven't reverted the merge and rebuilt for stretch to double check that it's not strictly speaking a Stretch issue. But Ken's findings above, suggest that squid isn't being appropriately patched so as to run as the `proxy` user.

Looking at that merge commit, I can see that changes were made to the way the code is patched. Specifically [this commit](https://github.com/turnkeylinux/tklbam-squid/commit/5c7c834c7a3dbf9cce877697b8a126394ee28206) jumps out at me, the commit message is "Migrate away from dpatch." The file in question; `debian/patches/01-cf.data.debian.dpatch` was renamed to `debian/patches/01-cf.data.debian.patch` and the execute bit removed (755 -> 644). I'm not clear on the significance of that.

It's possible that I'm missing something (highly likely due to the fact that I don't know Debian packaging that well, nor TKLBAM code and didn't do this troubleshooting) but it seems that that particular patch is no longer being applied at build time. So squid is running as it's default user `nobody`.

Do you want me to confirm that this issue is related to the reproducable packing changes (i.e. revert the merge and rebuild and test)? Or are you good to have a look at it?

**Comment by JedMeister** [comment]
@lamby - PS I've assigned you. If you'd like me to look into it first, please reassign back to me. Cheers.

**Comment by lamby** [comment]
@JedMeister dpatch is a weird patch system where the patches are executable. That commit attempts to move to the "quilt" Debian package format which is far more regular. There should be zero semantic changes whatsoever; the patches should be applied just as before just not in such a clanky way.

**Comment by JedMeister** [comment]
Thanks @lamby ok then, I'll have a closer look.

**Comment by JedMeister** [comment]
@lamby - finally got to having a look at this... Despite this:

> There should be zero semantic changes whatsoever; the patches should be applied just as before just not in such a clanky way.

It appears that the move from dpatch to quilt is indeed the issue. If I revert [5c7c834](https://github.com/turnkeylinux/tklbam-squid/commit/5c7c834c7a3dbf9cce877697b8a126394ee28206) and rebuild the package, squid (and tklbam) work fine. 

Does using dpatch interfere with the reproducability of the package? (I guess I could rebuild it and check, but figured it's quicker to ask). If not, I'm tempted to just run with it as is for now and try again later...? 

Thoughts?

**Comment by lamby** [comment]
> Does using dpatch interfere with the reproducability of the package?

(Yes, it would.)

> I'm tempted to just run with it as is for now and try again later...?

Hm, don't quite understand that. Or rather: why not just fix the problem as it stands instead of papering over problems when reverting..?

Anyway, see https://github.com/turnkeylinux/tklbam-squid/pull/1 for the commits that fix this.

**Comment by JedMeister** [comment]
Thanks mate. Agreed, fixing properly is (much, much) better, but I didn't actually expect you to be so incredibly responsive and fix it so quick!

**Comment by JedMeister** [comment]
@lamby @DocCyblade Just merged turnkeylinux/tklbam-squid#1, rebuilt the package(s) and tested a restore and it works! :tada: Woohoo! :smile:

**Comment by lamby** [comment]
Sweet!
